src/features/Coupon/utils/combinations.ts

```diff
@@ -0,0 +1,49 @@
+import { CartItem } from '@/features/Cart/types/Cart.types';
+import { Coupon } from '../types/Coupon.types';
+import { calculateTotalDiscount } from './calculateTotalDiscount';
+import { useCartInfo } from '@/features/Cart/hooks/useCartInfo';
+
+type PriceContext = {
+  isRemoteArea: boolean;
+  deliveryFee: number;
+  totalPrice: number;
+};
+
+export const getBestCouponCombination = (
+  coupons: Coupon[],
+  cartItems: CartItem[],
+  priceContext: PriceContext
+): Coupon[] => {
+  const { selectedCartItems } = useCartInfo(cartItems);
+  const combos = getAllCombinations(coupons, 2);
+  let bestCombo: Coupon[] = [];
+  let maxDiscount = 0;
+
+  for (const combo of combos) {
+    const discount = calculateTotalDiscount(selectedCartItems, combo, priceContext);
+    if (discount > maxDiscount) {
+      bestCombo = combo;
+      maxDiscount = discount;
+    }
+  }
+
+  return bestCombo;
+};
+
+const getAllCombinations = (arr: Coupon[], max: number) => {
+  const result: Coupon[][] = [];
+
+  const dfs = (start: number, path: Coupon[]) => {
+    if (path.length > 0 && path.length <= max) result.push([...path]);
+    if (path.length === max) return;
+
+    for (let i = start; i < arr.length; i++) {
+      path.push(arr[i]);
+      dfs(i + 1, path);
+      path.pop();
+    }
+  };
+
+  dfs(0, []);
+  return result;
+};
```

Comment:
탐색 알고리즘을 적용해주셨군요 ㅎㅎ 좋은 시도네요!

Reply:
이렇게 배열 메소드를 적극적으로 활용하는 방식도 고민해보시면 좋답니다!
```suggestion
const getAllCombinations = (arr: Coupon[], max: number): Coupon[][] => {
  const generateCombinations = (items: Coupon[], size: number): Coupon[][] => {
    if (size === 1) return items.map(item => [item]);

    return items.flatMap((item, index) =>
      generateCombinations(items.slice(index + 1), size - 1)
        .map(rest => [item].concat(rest))
    );
  };

  return Array.from({ length: max }, (_, i) => i + 1)
    .flatMap(size => generateCombinations(arr, size));
};
```

Reply:
새로운 방식 알아갑니다 !! 🫢